### ISSUE TITLE
Zookeeper gem fails to connect to valid zookeeper quorum

### ISSUE THREAD

**Issue by mlongob** [bloomberg/chef-bach#353]
This problem has catastrophic effects on HDFS-HA when one of the namenodes are down, as chef will bounce the active one.

```
def znode_exists?(znode_path, zk_host="localhost:2181")
  require 'rubygems'
  require 'zookeeper'
  znode_found = false
  begin
    zk = Zookeeper.new(zk_host)
    if !zk.connected?
      raise "znode_exists : Unable to connect to zookeeper quorum #{zk_host}"
    end 
    r = zk.get(:path => znode_path)
    if r[:rc] == 0
      znode_found = true
    end 
  rescue Exception => e
    puts e.message
  ensure
    if !zk.nil?
      zk.close unless zk.closed?
    end
  end
  return znode_found
end
```

This code fails on zk.connected? on a 3 node quorum where one of the node is unreachable

**Comment by mlongob** [comment]
chef >   zks = node[:bcpc][:hadoop][:zookeeper][:servers].map{|zkh| "#{zkh[:hostname]}:#{node[:bcpc][:hadoop][:zookeeper][:port]}"}.join(",")
chef > znode_exists?("/hadoop-ha/#{node.chef_environment}", zks)
 => true 
chef > znode_exists?("/hadoop-ha/#{node.chef_environment}", zks)
 => true 
chef > znode_exists?("/hadoop-ha/#{node.chef_environment}", zks)
 => true 
chef > znode_exists?("/hadoop-ha/#{node.chef_environment}", zks)
 => true 
chef > znode_exists?("/hadoop-ha/#{node.chef_environment}", zks)
 => true 
chef > znode_exists?("/hadoop-ha/#{node.chef_environment}", zks)
 => true 
chef > znode_exists?("/hadoop-ha/#{node.chef_environment}", zks)
znode_exists : Unable to connect to zookeeper quorum <QUORUM>
 => false 
chef > znode_exists?("/hadoop-ha/#{node.chef_environment}", zks)
 => true 
chef > znode_exists?("/hadoop-ha/#{node.chef_environment}", zks)
 => true 
chef > znode_exists?("/hadoop-ha/#{node.chef_environment}", zks)
 => true 
chef > znode_exists?("/hadoop-ha/#{node.chef_environment}", zks)
 => true 
chef > znode_exists?("/hadoop-ha/#{node.chef_environment}", zks)
 => true

**Comment by mlongob** [comment]
chef > z = Zookeeper.new(zks).connected?
 => true 
chef > z = Zookeeper.new(zks).connected?
 => true 
chef > z = Zookeeper.new(zks).connected?
 => true 
chef > z = Zookeeper.new(zks).connected?
 => true 
chef > z = Zookeeper.new(zks).connected?
 => true 
chef > z = Zookeeper.new(zks).connected?
 => false 
chef > z = Zookeeper.new(zks).connected?
 => false 
chef > z = Zookeeper.new(zks).connected?
 => true 
chef >

**Comment by mlongob** [comment]
Was able to reproduce with simple zookeper-client command. It looks like zookeeper-client chooses one node at random for kerberos authentication and does not retry on failiure. This is a potential critical bug. I'll investigate more tomorrow

**Comment by bijugs** [comment]
Looks like this is an [issue](https://issues.apache.org/jira/browse/ZOOKEEPER-1057) in ZK C client which got fixed in 3.4.1.
